Adding KubeVirt Cloud Provider

moadqassem opened the pull request:
**What this PR does / why we need it**:
Adding Kubevirt cloud provider


**Release note**:
```improvement operator
Adding kubevirt initial cloud provider 
```

moadqassem committed:
adding mc provider template

* adding tests for kubevirt cloud provider core module

* adding .ci dir

* injecting the kubevirt client in the cloud provider plugin

* add Dockerfile and build script

* fixing linter and vet issues

Signed-off-by: Moath Qasim <moad.qassem@gmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gardener/machine-controller-manager-provider-kubevirt?pullRequest=2) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/gardener/machine-controller-manager-provider-kubevirt?pullRequest=2) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gardener/machine-controller-manager-provider-kubevirt?pullRequest=2) it.</sub>

mfranczy reviewed (approved):
/lgtm
/approve

mvladev commented on cmd/machine-controller/main.go:
```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Can you update all the Copyright notices to 2020?

mvladev commented on go.mod:
```diff
@@ -0,0 +1,37 @@
+module github.com/moadqassem/machine-controller-manager-provider-kubevirt
```
`github.com/gardener/...`

mvladev commented on kubernetes/crds.yaml:
```diff
@@ -0,0 +1,357 @@
+    strategy: None
+  group: machine.sapcloud.io
+  names:
+    kind: AWSMachineClass
```
AWS?

moadqassem replied:
Yeah that was a copy paste mistake :-)

moadqassem replied:
Edit: 

In the crd file all the MachineClasses are included. The Kubevirt Machine Class can be found here: 
https://github.com/gardener/machine-controller-manager-provider-kubevirt/pull/2/files#diff-57ae387e5c536e1ade59b186d5dc2d1cR310

mvladev commented on pkg/kubevirt/core/core.go:
```diff
@@ -0,0 +1,300 @@
+			return "", fmt.Errorf("failed to create kubevirt client: %v", err)
+		}
+
+		if err := kubevirtClient.Delete(context.Background(), vm); err != nil {
```
nonfound error should be ignored in this case.

moadqassem replied:
Can you elaborate more about the meaning of "nonfound" error?

mvladev replied:
https://github.com/kubernetes/apimachinery/blob/7e441e0f246a2db6cf1855e4110892d1623a80cf/pkg/api/errors/errors.go#L478-L482

mvladev reviewed (dismissed):
Some changes are needed.

stoyanr reviewed (changes requested):
Thanks for the PR! Please see the comments on the specific lines of code, some of them apply to multiple files, please fix them everywhere. Also, please:
* Squash all commits into one (Gardener policy)
* Add a VERSION file
* Add `.ci` scripts from https://github.com/gardener/machine-controller-manager-provider-vsphere and make sure all Makefile targest work
* Do `make revendor`, currently some files are not properly vendored

moadqassem commented on Makefile:
```diff
@@ -0,0 +1,116 @@
+.PHONY: start
+start:
+	@GO111MODULE=on go run \
+			-mod=vendor \
```
This is actually generated from the script from the SampleProvider. Since most of these things are actually generated in the Makefile I left it as it is.

stoyanr replied:
I think it's redundant, so please remove it.

moadqassem commented on README.md:
```diff
@@ -1,3 +1,15 @@
+Out of tree (controller based) implementation for `kubevirt` as a new provider.
 
-To be done.
+## About
+- This is a sample repository that provides the blueprint required to implement a new (hyperscale) provider. We call the new provider as `kubevirt` for our ease.
+- Each provider implements the interface defined at [MCM OOT driver](https://github.com/gardener/machine-controller-manager/blob/master/pkg/util/provider/driver/driver.go).
+
+## Fundamental Design Principles:
+Following are the basic principles kept in mind while developing the external plugin.
+* Communication between this Machine Controller (MC) and Machine Controller Manager (MCM) is achieved using the Kubernetes native declarative approach.
+* Machine Controller (MC) behaves as the controller used to interact with the cloud provider and manage the VMs corresponding to the machine objects.
+* Machine Controller Manager (MCM) deals with higher level objects such as machine-set and machine-deployment objects.
+
+## Support for a new provider
+- Steps to be followed while implementing a new provider are mentioned [here](https://github.com/gardener/machine-controller-manager/blob/master/docs/development/cp_support_new.md)
```
Sure, I can do it in a different PR. For now I will remove this content and a TODO

moadqassem commented on go.mod:
```diff
@@ -0,0 +1,37 @@
+
+replace (
+	github.com/gardener/machine-controller-manager => github.com/prashanth26/machine-controller-manager v0.0.0-20200601182012-6f9dee78a746
+	github.com/onsi/gomega => github.com/onsi/gomega v1.5.0
```
This actually for the validation test for ginkgo. For now we can ignore this one.

stoyanr replied:
I think once you add `hack/tools.go` as mentioned below you should not need this line.

moadqassem commented on hack/rename-project:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash -eu
```
The whole `hack` folder should be gone since nothing is done from that side.

stoyanr replied:
In fact, I think you should add `tools.go` to your hack folder to make vendoring of test tools easier, see https://github.com/gardener/machine-controller-manager-provider-vsphere/blob/master/hack/tools.go

moadqassem commented on hack/rename-provider:
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash -eu
```
The whole `hack` folder should be gone since nothing is done from that side.

stoyanr replied:
See above.

moadqassem commented on kubernetes/crds.yaml:
```diff
@@ -0,0 +1,357 @@
+# The set of CRDs used by machine-controller-manager
```
Well, this controller runs as an OOT controller, I needed those crds to be applied to the test cluster, otherwise the controller will complain about missing crds. To create machines based on machine-classes per se you don't need to run the MCM and you can use this OOT controller alone. However, if you would like to create machine-deployment, then yes the MCM should be there. In other words, if someone would like to run this controller only( for testing e.g) then you need to apply those crds, and having them here would allow the tester to ignore MCM completely. 

Disclaimer: I have no problem of actually moving them, but having integration tests in the future for this project only would need those CRDS.

stoyanr replied:
I don't find them in the vsphere OOT controller, and I don't like the duplication much. I would prefer to have them removed and direct developers to their proper location if they are needed. When integration tests are added, we could decide to readd them back if this is the best option.

moadqassem replied:
I don't like duplication either tbh. As you said, let's get rid of them for now and we think about adding them later.

moadqassem commented on pkg/kubevirt/core/core.go:
```diff
@@ -0,0 +1,320 @@
+		},
+	}
+
+	if p.client == nil {
+		kubevirtClient, err := kubevirtClient(secrets)
+		if err != nil {
+			return "", fmt.Errorf("failed to get kubevirt client: %v", err)
+		}
+		p.client = kubevirtClient
+	}
```
This means you are actually relying on the constructor completely and any change in the secret wouldn't matter at all(only on the provider initialising). If this ok for your architecture then sure we can do that. But then passing those secrets, which to the methods won't make any sense(which is following the driver interface).

stoyanr replied:
In your existing code, if the secrets change, this will also not have any effect since you are saving the created client in the `client` field, so any subsequent calls with a different secret would use the already created client.
Regarding whether the secret can change: as far as I understand, when creating the client, you are only using the `kubeconfig` for your "provider" cluster from the secrets, correct? Can this "provider" cluster change during the lifetime of a shoot? I don't think so, but to be on the safe side, we should probably discuss it.
To be on the safe side, and having in mind that the secrets are passed in, my initial approach of injecting the client should be changed (and it's actually not feasible since without the secrets you can't create it, and at the point in time of calling the constructor you don't have the secrets, only during an actual method call). So therefore, you could inject a `ClientFactory` - an interface that has a single `NewClient(kubeconfig []byte)` method. The actual implementation of this factory would create the client. Note that you should not in this case save the result in the field, just use the created client only within the method it was created.

moadqassem commented on pkg/kubevirt/core/core.go:
```diff
@@ -0,0 +1,320 @@
+	}
+	if virtualMachine != nil {
+		virtualMachine.Spec.Running = utilpointer.BoolPtr(false)
+		if err := p.client.Update(ctx, virtualMachine); err != nil {
+			return "", fmt.Errorf("failed to update machine running state: %v", err)
+		}
```
But then you are actually handling the re-update inside the provider itself, where the requests to retry again from the reconciler itself not the cloud provider. Similar like what do you do in the machine creation. So when something goes wrong and re-enqueue the request again. Or do you have some certain criteria for re-enqueuing requests?

stoyanr replied:
As discussed in the meeting, both approaches are valid. I tend to slightly prefer handing `Conflict` errors in the way recommended above, since they are very common. This is also what's typically done in the Gardener codebase.

moadqassem commented on pkg/kubevirt/core/core_test.go:
```diff
@@ -0,0 +1,162 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
```
Why should I reinvent the wheel :-). This client is giving the tests all what it needs.

stoyanr replied:
This is not about reinventing the wheel, it's about a more comprehensive / universal approach to testing. Interface mocks are universal, can be generated for any interface (not just `client.Client`), and can be easily setup to imitate any behavior. If a mock is called with the wrong number / values of parameters, it's immediately an error. All this you don't get with the `fake` package. 
Note that this is also the approach used in much of the Gardener codebase, it should be used here as well. This can be done in a new PR as I mentioned above.

moadqassem replied:
Well, again, this universal approach can be valid of course if you have a type that needs to be mocked. But here you have one which is the fake client! a universal approach for dealing with k8s objects when testing. Another approach could be using ClientSet with Invokes however the case that I am testing doesn't need that layer of complexity. I am just creating, getting and deleting objects. If I need to react on something I can add da reactors in there, but this part of the rest client more than the controller runtime one! Otherwise mocking these clients is just a boilerplate as there is no special thing to do here!

moadqassem replied:
Anyway, if your codebase is assuming that each client has to be mocked explicitly then sure we can do that.

moadqassem replied:
So as agreed during the meeting sync, will update the test file in a separate PR.

mfranczy commented on Makefile:
```diff
@@ -0,0 +1,116 @@
+docker-image:
+	@docker build -t $(IMAGE_REPOSITORY):$(IMAGE_TAG) .
+
+.PHONY: docker-login
+docker-login:
+	@gcloud auth login
+
```
there is such a target in other extension providers (like vsphere), I would say it's quie convenient, for sure it doesn't harm to have it.

stoyanr replied:
OK, yes, you are right, but then please change it to match these other providers:
```
.PHONY: docker-login
docker-login:
	@gcloud auth activate-service-account --key-file .kube-secrets/gcr/gcr-readwrite.json
```

moadqassem commented on pkg/kubevirt/core/core.go:
```diff
@@ -0,0 +1,320 @@
+	return providerIDs, nil
+}
+
+func kubevirtClient(secret *corev1.Secret) (client.Client, error) {
```
Well, it doesn't have to be public there is a util file in the core package which can be moved to

stoyanr replied:
OK, that would be even better.

moadqassem commented on pkg/kubevirt/plugin.go:
```diff
@@ -0,0 +1,56 @@
+	corev1 "k8s.io/api/core/v1"
+)
+
+// PluginSPI provides an interface to deal with cloud provider session
```
This is all from the vSphere OOT controller!

stoyanr replied:
OK, I see, it's ok then.

stoyanr commented:
@moadqassem Could you also please add `install-requirements`, `format`, `check-generate`, `generate`,  `test-cov`, `test-clean`, `verify`, and `verify-extended` targets to the `Makefile`? As an example, see https://github.com/gardener/gardener-extension-provider-kubevirt/blob/master/Makefile. Please also change `check` and `test` target to use `/vendor/github.com/gardener/gardener/hack`. Then, when #4 is merged, we would get a `verify` check on future PRs.

stoyanr commented on pkg/kubevirt/core/core.go:
```diff
@@ -0,0 +1,283 @@
+}
+
+// NewPluginSPIImpl creates a new kubevirt cloud provider based on the passed client or secret.
+func NewPluginSPIImpl(client client.Client, secret *corev1.Secret) (*PluginSPIImpl, error) {
```
I don't see this method called anywhere except in tests? How is the client initialized in the production case? I don't think it is.
Also, I am not particularly fond of passing both a client and a secret, especially since you can create the client from the secret. I would remove the second parameter and make the client creation external to the constructor, or I would add a second constructor with a secret and call the first one from it.

moadqassem replied:
Yeah I still need to do some refactoring. The problem is, we followed the vSphere implementation, where in the previous one, we were creating the client from the secret which is passed to all methods that the constructor has. Now, this is not possible because the secret is not passed at the time where the client is created. And the client is created from the secret because it has the kubeconfig for the kiubevirt cluster!!

stoyanr reviewed (changes requested):
@moadqassem Thanks for addressing all remaining comments! Please see a new comment that I think is probably a bug. 
Reg. my comments about adding additional targets to the Makefile using Gardener `hack` scripts, that could also be addressed in a future PR.

moadqassem commented:
> @moadqassem Thanks for addressing all remaining comments! Please see a new comment that I think is probably a bug.
> Reg. my comments about adding additional targets to the Makefile using Gardener `hack` scripts, that could also be addressed in a future PR.

Yeah I will take care of that :-). I also still need to squash my comments and wait for another PR to be merged on the original repo(it is a PR that @mfranczy has opened regarding Dockerfile)

mfranczy commented:
@moadqassem the docker file PR has been rebased https://github.com/moadqassem/machine-controller-manager-provider-kubevirt-1/pull/1

stoyanr reviewed (changes requested):
Thanks for addressing the constructor issue! Some more requests:
* Please add +x flag on `.ci/build` and `.ci/check`.
* Please fix all linter issues:
```
$ make check
Running go vet...
# github.com/gardener/machine-controller-manager-provider-kubevirt/cmd/machine-controller
cmd/machine-controller/main.go:45:3: Fprint call has possible formatting directive %v
make: *** [check] Error 2
````
* Please add a `verify` target that calls `test` and `check`. We should use it for a `verify` check (until we add `verify-extended` as suggested previously).
